pkg/elf/elf.go

```diff
@@ -0,0 +1,260 @@
+	if len(intOptions) == 0 && len(strOptions) == 0 {
+		// Copy the remaining portion of the file
+		if _, err := io.Copy(w, r); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	globalOff := uint64(0) // current position in file
+	processedOptions := make(map[string]struct{}, len(intOptions)+len(strOptions))
+processSymbols:
+	for _, symbol := range elf.symbols.sort() {
+		scopedLog := log.WithField("symbol", symbol.name)
+
+		// Figure out the value to substitute
+		var value []byte
+		switch symbol.kind {
+		case symbolUint32:
+			v, exists := intOptions[symbol.name]
+			if exists {
+				value = make([]byte, unsafe.Sizeof(v))
+				elf.ByteOrder.PutUint32(value, v)
+			}
+		case symbolString:
+			v, exists := strOptions[symbol.name]
+			if exists {
+				if uint64(len(v)) != symbol.size {
+					return fmt.Errorf("symbol substitution value %q (len %d) must equal length of symbol name %q (len %d)", v, len(v), symbol.name, len(symbol.name))
+				}
+				value = []byte(v)
+			}
+		}
+		if value == nil {
+			for _, prefix := range ignoredPrefixes {
+				if strings.HasPrefix(symbol.name, prefix) {
+					continue processSymbols
+				}
+			}
+			scopedLog.Warning("Skipping symbol substitution")
+			continue processSymbols
+		}
+
+		// Copy data up until this symbol into the new file;
+		// Write the new value and seek past it.
+		dataToCopy := int64(symbol.offset - globalOff)
+		if _, err := io.CopyN(w, r, dataToCopy); err != nil {
+			return err
+		}
+		if err := binary.Write(w, elf.ByteOrder, value); err != nil {
+			return fmt.Errorf("failed to substitute %s: %s", symbol.name, err)
+		}
+		if _, err := r.Seek(int64(symbol.size), io.SeekCurrent); err != nil {
+			return err
+		}
+		processedOptions[symbol.name] = struct{}{}
+		globalOff = symbol.offset + symbol.size
+	}
+
+	// Check for additional options that weren't applied
+	for symbol := range strOptions {
+		if _, processed := processedOptions[symbol]; !processed {
+			return fmt.Errorf("no such string %q in ELF", symbol)
+		}
+	}
+	for symbol := range intOptions {
+		if _, processed := processedOptions[symbol]; !processed {
+			return fmt.Errorf("no such symbol %q in ELF", symbol)
+		}
+	}
+
+	// Copy the remaining portion of the file
+	if _, err := io.Copy(w, r); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Write the received ELF to a new file at the specified location, with the
+// specified options (indexed by name) substituted in the data section.
+// On success, writes the file to the specified path. On failure, returns an
+// error and no file is left on the filesystem.
+//
+// Concurrent writes from the same source ELF is NOT safe.
+func (elf *ELF) Write(path string, intOptions map[string]uint32, strOptions map[string]string) error {
+	f, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+
+	err = elf.copy(f, elf.file, intOptions, strOptions)
+	if err == nil {
+		err = f.Sync()
+	}
+	if _, err2 := elf.file.Seek(0, io.SeekStart); err2 != nil {
```

Comment:
FYI, since you are assigning a new variable you can still use `err` and the value set in line 231 won't be lost

Reply:
Nice to know, but this kind of accidental hiding is really easy to get wrong so I'd rather be a bit more explicit here.

Reply:
I ended up removing this anyway to make it more (likely to be) safe for concurrent use.